[FEA] Add option to enable diagnostic views generation in Profiler tools

cindyyuanjiang opened the pull request:
Fixes https://github.com/NVIDIA/spark-rapids-tools/issues/1620

This PR introduces a new command-line option to enable or disable diagnostic views generation in Profiling tool and removes diagnostic views generation from Qualification tools.

-------------------------------------------------------------------------------------------------------------------------

### Why:
- There are two Diagnostic views currently -> `StageDiagnosticMetrics`, `IODiagnosticMetrics`
- They rely on the aggregated data internally to render extra detail for developers in case they want to dig deep
- Since they are not needed in default scenarios, this change saves the extra memory and compute overhead of rendering these views
- On top of adding a param to control them, we also disable these by default from the `Qualification` context

### How:

This adds a new argument to the Application context for the profiler which then trickles downward to control if a diagnostic view is to be rendered or not

### Changes:

#### Diagnostic Views Feature:
* Added a new `enableDiagnosticViews` toggle in the `Profiler` class and `ProfileArgs` to allow users to enable or disable diagnostic views. This is disabled by default.
* Incorporated logic in `AppSQLPlanAnalyzer` and `AppSparkMetricsAnalyzer` to conditionally process and return diagnostic metrics only when the toggle is enabled.
* Updated the `ApplicationInfo` class to include an `isDiagnosticViewsEnabled` method for checking the diagnostic views toggle. 

#### Refactoring and Renaming:
* Renamed `ProfSparkMetricsAnalyzer` to `ProfSparkMetricsAggregator` and `QualSparkMetricsAnalyzer` to `QualSparkMetricsAggregator` for better alignment with their functionality. 

#### Test and Utility Updates:
* Modified `ToolTestUtils` to accept `enableDiagnosticViews` as a parameter for testing purposes. 
* Added new tests to AnalysisSuite and ApplicationInfoSuite

### Testing

Verified results from user_tools ->

#### Profiling tool:

`spark_rapids profiling -v -e <my-event-logs> -p onprem -t <jar-after-changes-in-this-PR>
`

`spark_rapids profiling -v -e <my-event-logs> -p onprem -t <jar-after-changes-in-this-PR> --enable-diagnostic-views
`

The two diagnostic files were not rendered by default in the Profiler

Tested to verify that Qualification does not render these views in the raw_metrics

-------------------------------------------------------------------------------------------------------------------------

### Performance Benchmarks

Event log size - `349MB` compressed
Memory overhead by adding the views ->
 IODiagnosticResult -> `6.5MB` shallow, `9.5MB` retained
 StageDiagnosticResult -> `5.5MB` shallow, `5.5MB` retained

This does **not** add a lot of stress.

For the runtime impact, the two views for the same log took a total time of `300ms` to be written.
For the calculation overhead, I barely saw any difference in terms of run time of the aggregator class ~ `50ms`.

amahussein reviewed (changes requested):
Thanks @cindyyuanjiang 
I have a couple of comments as it was not clear from the PR description what the changes are.

- Is this PR removing the views from the Qual context?
- The new argument seems to enable/diable the CSV files but it seems that the code that constructs those views has not changed. This implies that the overhead of constructing the views is still occurred in both Qual/Prof. Do I miss something?

cindyyuanjiang commented:
> Thanks @cindyyuanjiang I have a couple of comments as it was not clear from the PR description what the changes are.

Thanks @amahussein for catching this!

> * Is this PR removing the views from the Qual context?

Yes, currently in Qual tool, stage diagnostics(view 1) is generated. I believe we do not need this in Qual context so I removed stage diagnostics from Qual context.

> * The new argument seems to enable/diable the CSV files but it seems that the code that constructs those views has not changed. This implies that the overhead of constructing the views is still occurred in both Qual/Prof. Do I miss something?

Yes, you are correct. Sorry I missed this. I have updated the PR. I will also run some performance benchmarks to confirm this.

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/Profiler.scala:
```diff
@@ -396,11 +398,13 @@ class Profiler(hadoopConf: Configuration, appArgs: ProfileArgs, enablePB: Boolea
-    profileOutputWriter.writeCSVTable(ProfIODiagnosticMetricsView.getLabel,
-      profilerResult.diagnostics.IODiagnostics)
 
+    if (enableDiagnosticViews) {
```
Is there a theme that we follow while rendering labels ? Mainly why does the Stage_Diagnostic_label come from package object but the IODiagnosticMetrics come from a separate view ?

cindyyuanjiang replied:
AFAIK, there are 2 types of views:
1. Aggregated views, these come from [`RawMetricProfilerView.getAggMetrics`](https://github.com/NVIDIA/spark-rapids-tools/blob/dev/core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/Profiler.scala#L253). The labels for these views are defined in the package object.
2. All other views, these come from [`CollectInformation(analyzedApps)`](https://github.com/NVIDIA/spark-rapids-tools/blob/dev/core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/Profiler.scala#L226). The labels from these come from the view itself.

I tried to follow the existing convention. I think the reason is because, for the aggregated views (case 1), they are all stored in a single object `ProfilerAggregatedView`, while for the views in case 2, each view is an object, so it makes sense to define the label in each of the view's object.

sayedbilalbari reviewed (commented):
Thanks Cindy. Just some clarifying questions for my understanding.

cindyyuanjiang commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSparkMetricsAggTrait.scala:
```diff
@@ -29,13 +29,15 @@ trait AppSparkMetricsAggTrait extends AppIndexMapperTrait {
       index: Int = 1,
-      sqlAnalyzer: Option[AppSQLPlanAnalyzer] = None): AggRawMetricsResult = {
+      sqlAnalyzer: Option[AppSQLPlanAnalyzer] = None,
+      enableDiagnosticViews: Boolean = false): AggRawMetricsResult = {
```
Yes, that is correct. We want to disable these views by default.

cindyyuanjiang commented on core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ProfileArgs.scala:
```diff
@@ -112,6 +112,10 @@ Usage: java -cp rapids-4-spark-tools_2.12-<version>.jar:$SPARK_HOME/jars/*
         "that all workers are homogenous. It requires the AutoTuner to be enabled. Default is " +
         "./worker_info.yaml",
       default = Some(ProfilingAutoTunerConfigsProvider.DEFAULT_WORKER_INFO_PATH))
+  val enableDiagnosticViews: ScallopOption[Boolean] =
```
thanks @bilalbari! I see scallop support shorthand, I will fix this in the next iteration.

cindyyuanjiang replied:
After looking more into this, I don't see any other core tools options have shorthand support. I don't want to make this a special case yet. If okay, I will leave the option as it is now, but please comment if you have a strong opinion about this. Thanks.

cindyyuanjiang commented:
> Thanks @cindyyuanjiang ! The changes in the ProfArgs are fine in general. Regarding the changes in the analysis package, there is an alternative implementation that takes advantage of the OOP defined in the code. Instead of adding an argument to enable/disable the diagnostics view, we could have a new class that extends the current aggregator. This way we would have 2 clean implementations, one with Diagnostics and one without. The `object ProfSparkMetricsAnalyzer extends AppSparkMetricsAggTrait` would be the entry to make that split in implementation

Understood. Thanks @amahussein for the review!

> I can take over the task to add the classes to separate the implementations of Diagnostics view into their own .

Got it. How do we do this? Should I make changes to this PR and merge it?

cindyyuanjiang commented:
> * If you were able to generate numbers to see the performance impact of enabling Diagnostics views on your branch and we found that there is considerable difference, then we can: 1- move fwd with merging this PR without the suggested refactor; and 2- file a followup issue to refactor the classes as suggested in my comment.
> * If we proceed with refactoring plan: We can mark that as draft, then I can create a new PR building on top of the changes in ProfileArgs

I included some performance numbers in this PR description. Copied them here:

Running benchmarks on a relatively large event log after changes in this PR (3 runs):

Disabled diagnostic views
Best time: 318217 ms
Ave time: 386212 ms

Enabled diagnostic views
Best time: 348906 ms
Ave time: 354084 ms

Personally I don't think these numbers are very conclusive. I don't know if it is because:
1) this does not include changes in view 3
2) the event log is not large enough

WDYT? @amahussein

amahussein commented:
> Personally I don't think these numbers are very conclusive. I don't know if it is because:
this does not include changes in view 3
the event log is not large enough

It is not clear from the description above the eventlog size used in the test the JVM heap used to run the benchmarks.
Regardless, I see that the PR is not completely turning off work related to diagnostics. For example, the diagnostics hashMap is being updated in the process. This means that although the vews are not actually generated, the hashMap is still constructed in memory.

cindyyuanjiang commented:
> It is not clear from the description above the eventlog size used in the test the JVM heap used to run the benchmarks. Regardless, I see that the PR is not completely turning off work related to diagnostics. For example, the diagnostics hashMap is being updated in the process. This means that although the vews are not actually generated, the hashMap is still constructed in memory.

I have updated the PR to turn off collection of metrics for diagnostic views completely. If this looks good, then we can move onto performance benchmarks. Thanks @amahussein!

amahussein reviewed (commented):
Thanks @cindyyuanjiang 
I posted a question about `getAggRawMetrics`

amahussein commented:
Convert to draft until comments are addressed and conflicts are resolved.

cindyyuanjiang commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSparkMetricsAggTrait.scala:
```diff
@@ -46,19 +48,25 @@ trait AppSparkMetricsAggTrait extends AppIndexMapperTrait {
       analysisObj.aggregateDurationAndCPUTimeBySql(index),
       Seq(analysisObj.maxTaskInputSizeBytesPerSQL(index)),
-      analysisObj.aggregateDiagnosticMetricsByStage(index, sqlAnalyzer))
+      if (enableDiagnosticViews) {
+        analysisObj.aggregateDiagnosticMetricsByStage(index, sqlAnalyzer)
+      } else {
+        Seq.empty
+      })
```
Thank you @amahussein! Let me clarify this: 

In Qualification context, `sqlAnalyzer ` is passed to `aggregateDiagnosticMetricsByStage`. Since `sqlAnalyzer : class AppSQLPlanAnalyzer(app: AppBase, val enableDiagnosticViews: Boolean = false)`, we do not need to pass `enableDiagnosticViews` to `getAggRawMetrics`. I see this point.

In the Profiler context, `sqlAnalyzer=None` is not passed to `aggregateDiagnosticMetricsByStage`. In `getAggRawMetrics`, there is no way to see whether we should generate diagnostic views without passing the flag `enableDiagnosticViews` to the function.

Please let me know if I understand this wrong. cc: @sayedbilalbari

sayedbilalbari replied:
Hi Cindy, in my understanding, either we would need the sqlAnalyser in the profiler context or use the extra enableDiagnosticView param. Since sqlAnalyser is already enabled, let's use that.
Let's evaluate the trade off of using one, my preference is that since it already has the sqlAnalyser, pass that and use it to toggle diagnostic views. I will also take a look into the pros and cons of both and see structurally what makes sense.

amahussein commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/ProfSparkMetricsAggregator.scala:
```diff
@@ -20,6 +20,6 @@ package com.nvidia.spark.rapids.tool.analysis
 // extraction of AppIndex from ApplicationInfo. However, in the future this object can be used
 // to provide customized logic for the Profiler (i.e., handle metrics specific to GPU eventlogs)
-object ProfSparkMetricsAnalyzer extends AppSparkMetricsAggTrait with ProfAppIndexMapperTrait {
+object ProfSparkMetricsAggregator extends AppSparkMetricsAggTrait with ProfAppIndexMapperTrait {
```
Did we need to do that renaming? just wonder the motivatioon

amahussein commented on user_tools/src/spark_rapids_pytools/resources/profiling-conf.yaml:
```diff
@@ -37,6 +37,8 @@ sparkRapids:
       - d
       - driverlog
+      - e
+      - enable-diagnostic-views
```
Does the user-tools wrapper considered that those arguments are disabled by default as well?

sayedbilalbari replied:
Yes, diagnostic-views are disabled by default.

amahussein commented on user_tools/src/spark_rapids_pytools/resources/profiling-conf.yaml:
```diff
@@ -37,6 +37,8 @@ sparkRapids:
       - application-name
       - d
       - driverlog
+      - e
```
This `e` overlaps with the eventlogs shortcut of the user_tools cli.
Did we actually test those shortcuts would work fine?

sayedbilalbari replied:
Correctly pointed. I was going through the flow to check how the extra tools options are processed.
So this list whitelists the params that can be passed to the underlying java command without explicit validation or checks other than the whitelisting.
Updated the list to remove the `e` shorthand notation. Although using it would not fail execution, `e` being reserved by eventlogs is popped from the dictionary.
Only allowing the --enable-diagnostic-views to be whitelisted, and any other shorthand for this other than `e` does not have logical coherence.

amahussein reviewed (changes requested):
Thanks @sayedbilalbari and @cindyyuanjiang !
Some minor questions and comments.
I am kind of want to get done with this PR because I realized it keeps dragging and I almost forgot the context.

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/views/QualRawReportGenerator.scala:
```diff
@@ -100,14 +100,12 @@ object QualRawReportGenerator extends Logging {
       pWriter.writeText("\n### B. Analysis ###\n")
-      constructLabelsMaps(QualSparkMetricsAnalyzer.
+      constructLabelsMaps(QualSparkMetricsAggregator.
+        // In context of Qual, we don't need the Diagnostic Views
+        // That is handled because the QualAggregator returns an empty Seq
```
Yes, instead of using a toggle to do selective writing, the getAggRawMetrics returns an empty sequence in case of Qualification.
The overall logic breakdown for context setting as well ->
- We use two maps to maintain Diagnostic metrics ( IO and StageDiagnostics )
- In case of Qualification or Profiler with Diagnostic turned off, no data is rendered for these maps
- So they are essentially empty
- Previously we were adding another code change here that checked if Qual, then don't write
- But that was not needed since no data was added in case of Qual
- Hence the extra comment and removal of the extra checks

parthosa replied:
nit: Do we want to keep this comment? It is a very specific comment about the change in this PR and seems out of place if we look at from outside this PRs context.

sayedbilalbari replied:
Updated to remove the comment

parthosa commented on core/src/test/scala/com/nvidia/spark/rapids/tool/profiling/ApplicationInfoSuite.scala:
```diff
@@ -888,71 +889,75 @@ class ApplicationInfoSuite extends FunSuite with Logging {
-        val res = df.collect()
-        assert(res.nonEmpty)
-      }
+  test("test csv file output with and without diagnostic views") {
```
Can we have the cases as individual tests? This way it is easier to pinpoint which case failed

sayedbilalbari replied:
Each case has a testname which is present in the assert output, making it easy to identify which exact case has failed

parthosa replied:
But it would run as a single unit test. 

For different cases, we usually run them as separate unit tests. See - 
https://github.com/NVIDIA/spark-rapids-tools/blob/bebe567441e04c9efb1ee352851580da9f8200ad/core/src/test/scala/com/nvidia/spark/rapids/tool/planparser/SqlPlanParserSuite.scala#L1232

sayedbilalbari commented on core/src/test/scala/com/nvidia/spark/rapids/tool/ToolTestUtils.scala:
```diff
@@ -146,7 +146,8 @@ object ToolTestUtils extends Logging {
   def processProfileApps(logs: Array[String],
-      sparkSession: SparkSession): ArrayBuffer[ApplicationInfo] = {
+      sparkSession: SparkSession,
+      enableDiagnosticViews: Boolean = false): ArrayBuffer[ApplicationInfo] = {
```
Updated to use enable diagnostic views using enableDiagnosticViews args from ProfileArgs

cindyyuanjiang commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSparkMetricsAnalyzer.scala:
```diff
@@ -310,12 +310,23 @@ class AppSparkMetricsAnalyzer(app: AppBase) extends AppAnalysisBase(app) {
   def aggregateDiagnosticMetricsByStage(index: Int, analyzer: Option[AppSQLPlanAnalyzer] = None):
       Seq[StageDiagnosticResult] = {
+    // We only return the diagnostic metrics if the diagnostic views are enabled.
+    // This is done only in case of Profiler and that too when explicitly enabled.
```
nit: we can also align the comment here and in `AppSQLPlanAnalyzer`

sayedbilalbari replied:
Thank you for pointing out, updated

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/views/QualRawReportGenerator.scala:
```diff
@@ -100,14 +100,10 @@ object QualRawReportGenerator extends Logging {
-          } else {
-            pWriter.writeTable(label, metrics, AGG_DESCRIPTION.get(label))
-          }
+          pWriter.writeTable(label, metrics, AGG_DESCRIPTION.get(label))
```
In case of qual tool, since the diagnostic metrics are disabled, they are not even updated and this returns an empty Seq, hence no file is generated

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSparkMetricsAnalyzer.scala:
```diff
@@ -310,12 +310,23 @@ class AppSparkMetricsAnalyzer(app: AppBase) extends AppAnalysisBase(app) {
       Seq[StageDiagnosticResult] = {
+    // We only return the diagnostic metrics if the diagnostic views are enabled.
+    // This is done only in case of Profiler and that too when explicitly enabled.
+    val isDiagnosticEnabled = app match {
```
Well pointed, updated !

cindyyuanjiang commented:
Thanks @sayedbilalbari! LGTM.

amahussein reviewed (approved):
Thanks @sayedbilalbari !
LGTME
